File: python/forth/forth.py
from itertools import takewhile

L = list[int]

OPS = {
    "+": lambda stack: stack.append(stack.pop() + stack.pop()),
    "-": lambda stack: stack.append(stack.pop(-2) - stack.pop()),
    "*": lambda stack: stack.append(stack.pop() * stack.pop()),
    "/": lambda stack: stack.append(stack.pop(-2) // stack.pop()),
    "dup": lambda stack: stack.append(stack[-1]),
    "drop": lambda stack: stack.pop(),
    "swap": lambda stack: stack.extend([stack.pop(), stack.pop()]),
    "over": lambda stack: stack.append(stack[-2]),
}


class StackUnderflowError(Exception):
    pass


def evaluate(input_data: L, stack: L | None = None, user_defs: dict[str, L] | None = None) -> L:
    stack, user_defs = stack or [], user_defs or {}
    word_iter = iter(" ".join(input_data).lower().split())
    while word := next(word_iter, None):
        if word == ":":
            name = next(word_iter)
            if name.lstrip("-").isdigit():
                raise ValueError("illegal operation")

            old_instructions = user_defs.get(name)
            user_defs[name] = list(takewhile(lambda word: word != ";", word_iter))
            if old_instructions:
                for name2, instructions in user_defs.items():
                    user_defs[name2] = sum((old_instructions if x == name else [x] for x in instructions), [])
        elif word in user_defs:
            stack = evaluate(user_defs[word], stack, user_defs)
        elif word in OPS:
            try:
                OPS[word](stack)
            except IndexError as exc:
                raise StackUnderflowError("Insufficient number of items in stack") from exc
            except ZeroDivisionError as exc:
                raise ZeroDivisionError("divide by zero") from exc
        elif word.lstrip("-").isdigit():
            stack.append(int(word))
        else:
            raise ValueError("undefined operation")

    return stack

File: javascript/forth/forth.js
export class Forth {
  #stack = [];
  #defs = {};

  static OPS = {
    '+': {min_stack: 2, func: (f) => f.#stack.push(f.pop() + f.pop())},
    '-': {min_stack: 2, func: (f) => f.#stack.push(f.pop(2) - f.pop())},
    '*': {min_stack: 2, func: (f) => f.#stack.push(f.pop() * f.pop())},
    '/': {min_stack: 2, func: (f) => f.#stack.push(div(f.pop(2), f.pop()))},
    'dup': {min_stack: 1, func: (f) => f.#stack.push(f.peek(1))},
    'drop': {min_stack: 1, func: (f) => f.pop()},
    'swap': {min_stack: 2, func: (f) => f.#stack.push.apply(f.#stack, [f.pop(), f.pop()])},
    'over': {min_stack: 2, func: (f) => f.#stack.push(f.peek(2))},
  };

  /**
   * @param {string} commands
   * @throws {Error}
   */
  evaluate(commands) {
    let words = commands.split(' ').map((word) => word.toLowerCase()), iNext, word;
    for (let i = 0; i < words.length;) {
      if ((word = words[i++]) == ':') {
        throwIf((iNext = words.indexOf(';', i)) < (i + 1) || /^-?\d+$/.test(words[i]), 'Invalid definition');
        this.#defs[words[i]] = words.slice(i + 1, iNext).flatMap((w) => this.#defs[w] ?? w);
        i = iNext + 1;
      } else if (word in this.#defs) {
        this.evaluate(this.#defs[word].join(' '));
      } else if (word in Forth.OPS) {
        throwIf(this.#stack.length < 1, 'Stack empty');
        throwIf(this.#stack.length < 2 && Forth.OPS[word].min_stack == 2, 'Only one value on the stack')
        Forth.OPS[word].func(this);
      } else {
        throwIf(!/^-?\d+$/.test(word), 'Unknown command', () => this.#stack.push(Number(word)));
      }
    }
  }

  /**
   * @returns {number[]}
   */
  get stack() { return this.#stack; }

  /**
   * @param {number} n
   * @returns {number}
   * @throws {Error}
   */
  pop = (n = 1) => this.#stack.splice(-n, 1)[0];

  /**
   * @param {number} n
   * @returns {number}
   */
  peek = (n) => this.#stack[this.#stack.length - n];
}

/**
 * @param {boolean} cond
 * @param {string} msg
 * @param {function|null} func
 * @returns {number|undefined}
 * @throws {Error}
 */
const throwIf = (cond, msg, func = null) => { if (cond) { throw new Error(msg); } else { return func?.(); } }

/**
 * @param {number} a
 * @param {number} b
 * @returns {number}
 * @throws Error
 */
const div = (a, b) => throwIf(b == 0, 'Division by zero', () => Math.floor(a / b));
